### PR TITLE
Improve reported RPC submit errors (RIPD-600)

### PR DESCRIPTION
**Pull request by miguelportilla**

### CODE REVIEW

File: src/ripple/protocol/impl/STLedgerEntry.cpp [hunk 1]
```diff
@@ -76,15 +76,19 @@ void STLedgerEntry::setSLEType ()
         Throw<std::runtime_error> ("invalid ledger entry type");
 
     type_ = mFormat->getType ();
-    if (!setType (mFormat->elements))
+    try
     {
-        if (auto j = debugJournal().warn())
+        setType(mFormat->elements);
+    }
+    catch(std::exception&)
```
Comment (vinniefalco) [inline src/ripple/protocol/impl/STLedgerEntry.cpp]: Hmm we should not be using exceptions to indicate errors arising from user inputs. To be clear, these are examples of allowed uses of exceptions:
* A required file is unavailable
* An internal logic error is detected
* The configuration file has a syntax error

These are examples where throwing an exception should be avoided:
* A network message is invalid
* A client submits an object that cannot be deserialized

What the disallowed cases have in common, is the principle that untrusted inputs should not be allowed to make the program generate exceptions at-will.

File: src/ripple/protocol/impl/STObject.cpp [hunk 2]
```diff
@@ -116,7 +116,8 @@ bool STObject::setType (const SOTemplate& type)
                 JLOG (debugJournal().warn())
                     << "setType(" << getFName().getName()
                     << "): explicit default " << e->e_field.fieldName;
-                valid = false;
+                Throw<std::runtime_error>("Explicit default " +
+                    e->e_field.fieldName);
```
Comment (vinniefalco) [inline src/ripple/protocol/impl/STObject.cpp]: Same thing here, we took a function that uses to return a `bool` indicating success and now throws an exception based on untrusted network inputs.

File: src/ripple/protocol/impl/STLedgerEntry.cpp (hunk 1 above)
Comment (miguelportilla) [inline src/ripple/protocol/impl/STLedgerEntry.cpp]: Just to be clear, the existing code throws exceptions on invalid inputs. But I can change that.

File: src/ripple/protocol/impl/STObject.cpp (hunk 2 above)
Comment (miguelportilla) [inline src/ripple/protocol/impl/STObject.cpp]: True, but the returned false bool causes exceptions to be thrown.

Comment (vinniefalco) [inline src/ripple/protocol/impl/STObject.cpp]: Thats unfortunate, and I understand if your position is that changing that is beyond the scope of this set of changes. However, we should not build on that mistake.

File: src/ripple/protocol/impl/STObject.cpp [hunk 3]
```diff
@@ -141,13 +143,13 @@ bool STObject::setType (const SOTemplate& type)
             JLOG (debugJournal().warn())
                 << "setType(" << getFName().getName()
                 << "): non-discardable leftover " << e->getFName().getName ();
-            valid = false;
+            Throw<std::runtime_error>("non-discardable leftover " +
```
Comment (sublimator) [inline src/ripple/protocol/impl/STObject.cpp]: `Extraneous` ?

File: src/ripple/protocol/impl/STObject.cpp (hunk 2 above)
Comment (sublimator) [inline src/ripple/protocol/impl/STObject.cpp]: What do you suggest then Vinnie? Something like returning a `std::pair<bool, std::string>`?

Comment (vinniefalco) [inline src/ripple/protocol/impl/STObject.cpp]:
```
struct parse_error { ... };

void parse(parse_error& pe, ... );
```

**Pull request closed without merging**